### PR TITLE
TBB block parallel back-end

### PR DESCRIPTION
**Pull request by erikzenker**
This PR introduces a new alpaka back-end based on Intel TBB. It should be seen as a first draft using TBB in alpaka and follows the need specified in issue #65. The back-end is very similar to the OpenMP 2.0 block parallel back-end. Therefore, it supports blocks with one thread.

### CODE REVIEW

**Comment by BenjaminW3** [comment]
Thanks for adding this back-end!

**Comment by BenjaminW3** [comment]
Could you please add it to the configuration of the CI tests, at line 64 and 73 of the `travis.yml` file (similar to the surrounding lines), line 61 of the `appveyor.yml` and line 41 of `travis/compileExec.sh`?
Furthermore, the CI tests are green as the TBB back-end is not currently used. You will have to install TBB at the end of the `install` step of the `travis.yml` file.

File: include/alpaka/acc/AccCpuTbbBlocks.hpp [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+                TSize const & blockSharedMemDynSizeBytes) :
+                    workdiv::WorkDivMembers<TDim, TSize>(workDiv),
+                    idx::gb::IdxGbRef<TDim, TSize>(m_gridBlockIdx),
+                    idx::bt::IdxBtZero<TDim, TSize>(),
+                    atomic::AtomicStlLock(),
+                    math::MathStl(),
+                    block::shared::dyn::BlockSharedMemDynBoostAlignedAlloc(static_cast<std::size_t>(blockSharedMemDynSizeBytes)),
+                    block::shared::st::BlockSharedMemStMasterSync(
+                        [this](){block::sync::syncBlockThreads(*this);},
+                        // TODO replace me by correct value
```
Comment (BenjaminW3) [inline include/alpaka/acc/AccCpuTbbBlocks.hpp]: What is the correct value? ;-)

File: include/alpaka/exec/ExecCpuTbbBlocks.hpp [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+                                    std::ref(m_kernelFnObj),
+                                    std::placeholders::_1,
+                                    std::ref(args)...);
+                        },
+                        m_args));
+
+
+
+                // The number of threads in this block.
+                //TSize const blockThreadCount(blockThreadExtent.prod());
```
Comment (BenjaminW3) [inline include/alpaka/exec/ExecCpuTbbBlocks.hpp]: Could you please remove those lines if they are unused?

File: include/alpaka/exec/ExecCpuTbbBlocks.hpp [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+                                 );
+
+                             boundKernelFnObj(acc);
+
+                             block::shared::st::freeMem(acc);
+
+                         });
+
+                         g.wait();
+                         //block::sync::syncBlockThreads(acc);
```
Comment (BenjaminW3) [inline include/alpaka/exec/ExecCpuTbbBlocks.hpp]: Could you please remove this line if they are unused?

Comment (BenjaminW3) [inline include/alpaka/exec/ExecCpuTbbBlocks.hpp]: Could you please remove this line if it is useless?

File: include/alpaka/exec/ExecCpuTbbBlocks.hpp [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+                                      gridBlockExtent
+                                 );
+
+                             boundKernelFnObj(acc);
+
+                             block::shared::st::freeMem(acc);
+
+                         });
+
+                         g.wait();
```
Comment (BenjaminW3) [inline include/alpaka/exec/ExecCpuTbbBlocks.hpp]: Is it correct to wait here? Shouldn`t it be outside the loop to execute all blocks in parallel and wait for all of them to be finished?

File: include/alpaka/acc/AccCpuTbbBlocks.hpp (hunk 1 above)
Comment (erikzenker) [inline include/alpaka/acc/AccCpuTbbBlocks.hpp]: Wrong, concept. BlockSharedMemStNoSync should be used there. thx !

File: include/alpaka/exec/ExecCpuTbbBlocks.hpp (hunk 2 above)
Comment (erikzenker) [inline include/alpaka/exec/ExecCpuTbbBlocks.hpp]: Yes

File: include/alpaka/exec/ExecCpuTbbBlocks.hpp (hunk 3 above)
Comment (erikzenker) [inline include/alpaka/exec/ExecCpuTbbBlocks.hpp]: Yes

File: include/alpaka/exec/ExecCpuTbbBlocks.hpp (hunk 4 above)
Comment (erikzenker) [inline include/alpaka/exec/ExecCpuTbbBlocks.hpp]: I think it is correct, but I found a much more simpler way to describe the parallel code here ([tbb::parallel_for](https://software.intel.com/en-us/blogs/2009/08/03/parallel_for-is-easier-with-lambdas-intel-threading-building-blocks)).

**Comment by BenjaminW3** [comment]
github seems to be faulty at the moment as all builds currently fail with:
```
Cloning into 'libs/local_function'...
fatal: unable to access 'https://github.com/boostorg/local_function.git/': transfer closed with outstanding read data remaining
fatal: clone of 'https://github.com/boostorg/local_function.git' into submodule path 'libs/local_function' failed
```

**Comment by BenjaminW3** [comment]
Ah, I missed one more place to add the new back-end to: the ReadMe has two tables where it has to be added to as well as the dependencies section.
Please try to install TBB in the `travis.yml` and `appveyor.yml` CI scripts as it is not available by default on the build hosts. I may have a look at the windows part of it at the end of the week.

**Comment by bussmann** [comment]
I like that you always include the Windows world...

**Comment by BenjaminW3** [comment]
I have just looked into the build ouput. CMake isn`t even finding a `FindTBB.cmake` script. Do you have one installed locally on your developer machine? Do you intend to add one locally to `alpaka/cmake`? [This one](https://github.com/justusc/FindTBB) seems to be good (License, Platforms).

File: alpakaConfig.cmake [hunk 5]
```diff
@@ -288,6 +290,20 @@ IF(ALPAKA_ACC_GPU_CUDA_ENABLE)
 ENDIF()
 
 #-------------------------------------------------------------------------------
+# Find TBB.
+#-------------------------------------------------------------------------------
+IF(ALPAKA_ACC_CPU_B_TBB_T_SEQ_ENABLE)
+    FIND_PACKAGE(TBB)
+    IF(NOT TBB_FOUND)
+        MESSAGE(WARNING "Optional alpaka dependency TBB could not be found! TBB grdi block accelerator disabled!")
```
Comment (BenjaminW3) [inline alpakaConfig.cmake]: ~~grdi~~ grid

**Comment by erikzenker** [comment]
You are right. I was using a local `FindTBB.cmake` script from the same source that you mentioned. I will just add it to the particular Alpaka cmake folder.

**Comment by erikzenker** [comment]
The parallel_for expression was introduced at version 2.2, therefore, version 2.2 should be our minimum requirement of TBB. I will add this to cmake.

**Comment by erikzenker** [comment]
I have no idea if this windows TBB installation will work ...

**Comment by psychocoderHPC** [comment]
@erikzenker is this branch feature complete else we should add the WIP lable to this pull request.

**Comment by BenjaminW3** [comment]
You should probably move the find script to a new folder `cmake/modules/FindTBB.cmake` to follow CMake standards and add `cmake/modules` to the `CMAKE_MODULE_PATH` to make the script found.

File: appveyor.yml [hunk 6]
```diff
@@ -168,19 +169,27 @@ before_build:
     - cmd: b2 -j2 --toolset=msvc-14.0 --layout=versioned %ALPAKA_BOOST_B2% architecture=x86 address-model=%BOOST_ADDRESS_MODEL% variant=%BOOST_VARIANT% link=static threading=multi runtime-link=shared define=_CRT_NONSTDC_NO_DEPRECATE define=_CRT_SECURE_NO_DEPRECATE define=_SCL_SECURE_NO_DEPRECAT define=BOOST_USE_WINFIBERS --stagedir="%ALPAKA_B2_STAGE_DIR%"
 
     #-------------------------------------------------------------------------------
+    # Install TBB
+    - cmd: set TBB_ROOT_DIR=C:\projects\tbb
+    - cmd: git clone -b master --recursive --single-branch --depth 1 https://github.com/intel-tbb/intel-tbb.git %TBB_ROOT_DIR%
```
Comment (BenjaminW3) [inline appveyor.yml]: This clone will not be enough, as the prebuild dynamic libraries will not be installed.
You can remove this last line and I can add the correct installation later (not necessarily in this Pull Request).

**Comment by erikzenker** [comment]
@psychocoderHPC it is feature complete

**Comment by BenjaminW3** [comment]
Thanks, this is looking really good. I can merge it after you squashed the commits!

**Comment by psychocoderHPC** [comment]
Please wait with merging up to #222 is checked.

**Comment by erikzenker** [comment]
Yes, all copy/paste errors, thank you for the hint.

**Comment by erikzenker** [comment]
@psychocoderHPC #222 can be fixed independently of this PR. If the interface of 'AtomicStlLock' stays the same, then I don't see a reason to wait for #223.

File: include/alpaka/exec/ExecCpuTbbBlocks.hpp [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+
+                // The number of blocks in the grid.
+                TSize const numBlocksInGrid(gridBlockExtent.prod());
+
+                // There is only ever one thread in a block in the TBB accelerator.
+                assert(blockThreadExtent.prod() == 1u);
+
+                tbb::parallel_for(TSize(0), TSize(numBlocksInGrid), [&](TSize i){
+
+                             acc::AccCpuTbbBlocks<TDim, TSize> acc(
```
Comment (BenjaminW3) [inline include/alpaka/exec/ExecCpuTbbBlocks.hpp]: This block is indented too much.
I would prefer something like:
```C++
tbb::parallel_for(
    static_cast<TSize>(0),
    static_cast<TSize>(numBlocksInGrid),
    [&](TSize i){
        acc::AccCpuTbbBlocks<TDim, TSize> acc(
        ....
```

**Comment by BenjaminW3** [comment]
I agree with @erikzenker. As soon as the style issue is fixed, I would like to merge this.

**Comment by BenjaminW3** [comment]
@erikzenker #223 is merged now, so there is only the style issue left ;-)

**Comment by erikzenker** [comment]
@BenjaminW3 I fixed the style issue

**Comment by BenjaminW3** [comment]
Due to the merges to develop, this branch has to be rebased before it can be merged.

**Pull request merged**